tensorpack/dataflow/serialize.py

```diff
@@ -0,0 +1,191 @@
+        db = lmdb.open(path, subdir=isdir,
+                       map_size=1099511627776 * 2, readonly=False,
+                       meminit=False, map_async=True)    # need sync() at the end
+        size = _reset_df_and_get_size(df)
+        with get_tqdm(total=size) as pbar:
+            idx = -1
+
+            # LMDB transaction is not exception-safe!
+            # although it has a context manager interface
+            txn = db.begin(write=True)
+            for idx, dp in enumerate(df.get_data()):
+                txn.put(u'{:08}'.format(idx).encode('ascii'), dumps(dp))
+                pbar.update()
+                if (idx + 1) % write_frequency == 0:
+                    txn.commit()
+                    txn = db.begin(write=True)
+            txn.commit()
+
+            keys = [u'{:08}'.format(k).encode('ascii') for k in range(idx + 1)]
+            with db.begin(write=True) as txn:
+                txn.put(b'__keys__', dumps(keys))
+
+            logger.info("Flushing database ...")
+            db.sync()
+        db.close()
+
+    @staticmethod
+    def load(path, shuffle=True):
+        """
+        Note:
+            If you found deserialization being the bottleneck, you can use :class:`LMDBData` as the reader
+            and run deserialization as a mapper in parallel.
+        """
+        df = LMDBData(path, shuffle=shuffle)
+        return MapData(df, lambda dp: loads(dp[1]))
+
+
+class NumpySerializer():
+    """
+    Serialize the entire dataflow to a npz dict.
+    Note that this would have to store the entire dataflow in memory,
+    and is also >10x slower than the other serializers.
+    """
+    @staticmethod
+    def save(df, path):
+        """
+        Args:
+            df (DataFlow): the DataFlow to serialize.
+            path (str): output npz file.
+        """
+        buffer = []
+        size = _reset_df_and_get_size(df)
+        with get_tqdm(total=size) as pbar:
+            for dp in df.get_data():
+                buffer.append(dp)
+                pbar.update()
+        np.savez_compressed(path, buffer=buffer)
+
+    @staticmethod
+    def load(path, shuffle=True):
+        buffer = np.load(path)['buffer']
+        return DataFromList(buffer, shuffle=shuffle)
+
+
+class TFRecordSerializer():
+    """
+    Serialize datapoints to bytes (by tensorpack's default serializer) and write to a TFRecord file.
+
+    Note that TFRecord does not support random access and is in fact not very performant.
+    It's better to use :class:`LMDBSerializer`.
+    """
+    @staticmethod
+    def save(df, path):
+        """
+        Args:
+            df (DataFlow): the DataFlow to serialize.
+            path (str): output tfrecord file.
+        """
+        size = _reset_df_and_get_size(df)
+        with tf.python_io.TFRecordWriter(path) as writer, get_tqdm(total=size) as pbar:
+            for dp in df.get_data():
+                writer.write(dumps(dp).to_pybytes())
```

Comment:
This breaks legacy implementations based on `msgpack`. Is `msgpack` still supported?

Reply:
eventually no. But I should keep msgpack working.